Feature/vplay 8907 guard against thread overwrite

p-bond opened the pull request.

p-bond committed:
VPLAY-9677 Guard against fragmentCollectorThread being overwritten

Attempting to assign a thread to an existing running thread calls std::terminate()
Check joinable status and only create if the thread is not already running.

Remove boolean fragmentCollectorThreadStarted from FragmentCollector classes, as it is not needed anymore;
this is an intrinsic property of std::thread.

Testing:  L1 tests run (but investigating extending)

p-bond committed:
VPLAY-9677 Guard against fragmentCollectorThread being overwritten
Attempting to assign a thread to an existing running thread calls std::terminate()
Check joinable status and only create if the thread is not already running.

Remove boolean fragmentCollectorThreadStarted from FragmentCollector classes, as it is not needed anymore;
this is an intrinsic property of std::thread.

Testing:  L1 tests run (but investigating extending)

p-bond committed:
VPLAY-9677 Guard against fragmentCollectorThread being overwritten
Attempting to assign a thread to an existing running thread calls std::terminate()
Check joinable status and only create if the thread is not already running.

Remove boolean fragmentCollectorThreadStarted from FragmentCollector classes, as it is not needed anymore;
this is an intrinsic property of std::thread.

Testing:  L1 tests run (but investigating extending)

p-bond committed:
VPLAY-9677 Guard against fragmentCollectorThread being overwritten
Attempting to assign a thread to an existing running thread calls std::terminate()
Check joinable status and only create if the thread is not already running.

Remove boolean fragmentCollectorThreadStarted from FragmentCollector classes, as it is not needed anymore;
this is an intrinsic property of std::thread.

Testing:  L1 tests run (but investigating extending)

p-bond committed:
VPLAY-9677 Guard against fragmentCollectorThread  & TSBreaderThread being overwritten
Attempting to assign a thread to an existing running thread calls std::terminate()
Check joinable status and only create if the thread is not already running.

Remove boolean fragmentCollectorThreadStarted from FragmentCollector classes, as it is not needed anymore;
this is an intrinsic property of std::thread.

Testing:  L1 tests run (but investigating extending)

p-bond committed:
VPLAY-9677 Guard against fragmentCollectorThread  & TSBreaderThread being overwritten
Attempting to assign a thread to an existing running thread calls std::terminate()
Check joinable status and only create if the thread is not already running.

Remove boolean fragmentCollectorThreadStarted from FragmentCollector classes, as it is not needed anymore;
this is an intrinsic property of std::thread.

Testing:  L1 tests run (but investigating extending)

jfagunde commented on test/utests/tests/fragmentcollector_mpd/fragmentcollector_mpd1.cpp:
```diff
@@ -80,6 +80,38 @@ class MpdTests : public ::testing::Test
 			mCurrentPeriod = mpd->GetPeriods().at(0);
 		}
+
+		void TsbReader() override
```
This is overriding StreamAbstractionAAMP_MPD::TsbReader(), part of the code that is being tested. Why not use the real code?

p-bond replied:
The code being changed is to start the thread and is not testing the operation of the thread.

jfagunde commented on test/utests/tests/fragmentcollector_mpd/fragmentcollector_mpd1.cpp:
```diff
@@ -335,3 +367,40 @@ TEST_F(MpdTests, FindPositionInTimeline2)
+	// InitializeMPD("");
+	// Set local TSB to true
+	mStreamAbstractionAAMP_MPD = new TestableStreamAbstractionAAMP_MPD(mPrivateInstanceAAMP);
+	EXPECT_CALL(*g_mockPrivateInstanceAAMP, DownloadsAreEnabled()).WillRepeatedly(Return(false));
```
Why are downloads disabled for this test? The problematic scenario that we are facing is with AAMP TSB injection false and downloads enabled. The crash was happening because the FetcherLoop thread was started while it was already running.

jfagunde commented on test/utests/tests/fragmentcollector_progressiveTests/fragmentcollector_progressiveTests.cpp:
```diff
@@ -49,6 +49,21 @@ class fragmentcollector_progressiveTests : public ::testing::Test
+    // Assert that the thread ID is still valid (not zero)
+    ASSERT_NE(profileEvent->fragmentCollectorThreadID, std::thread::id());
+    // Assert that the thread ID has not changed
+    ASSERT_EQ(profileEvent->fragmentCollectorThreadID, std::this_thread::get_id());
```
I don't understand this check. Why is the thread ID of the fragment collector thread the same as the thread ID of the thread running the test?

p-bond commented on test/utests/tests/fragmentcollector_mpd/fragmentcollector_mpd1.cpp:
```diff
@@ -335,3 +367,40 @@ TEST_F(MpdTests, FindPositionInTimeline2)
+
+TEST_F(MpdTests, testRepeatedStartLocalTSB)
+{
+	// InitializeMPD("");
```
Done.

p-bond commented on fragmentcollector_mpd.cpp:
```diff
@@ -10471,11 +10471,19 @@ void StreamAbstractionAAMP_MPD::StartFromOtherThanAampLocalTsb(void)
+		}
+		else
+		{
+			AAMPLOG_WARN("FetcherLoop thread already running, not creating a new one");
```
Done.

p-bond committed:
Dependency injection added for testing of thread starting
Fakes & mocks extended to cope with this
Incidental change to related tests using strict mocks

p-bond committed:
Merge branch 'dev_sprint_25_2' into feature/VPLAY-8907-guardAgainstThreadOverwrite

cpc005 commented on fragmentcollector_hls.cpp:
```diff
@@ -1514,41 +1514,41 @@ bool TrackState::FetchFragmentHelper(int &http_error, bool &decryption_error, bo
 	if(ISCONFIGSET(eAAMPConfig_EnablePTSReStamp) && (aamp->mVideoFormat == FORMAT_ISO_BMFF ))
-    {
-    	if(playContext)
+	{
```
Changing space to tab in files not otherwise changed for this jira is unnecessary. It will overwrite the commit history of the lines

DomSyna commented on fragmentcollector_hls.cpp:
```diff
@@ -1514,41 +1514,41 @@ bool TrackState::FetchFragmentHelper(int &http_error, bool &decryption_error, bo
 		{
 			AAMPLOG_WARN(" %s abort waiting for video PTS arrival",name );
-          	playContext->abortWaitForVideoPTS();
+			  playContext->abortWaitForVideoPTS();
```
alignment

DomSyna commented on fragmentcollector_hls.cpp:
```diff
@@ -5073,17 +5071,24 @@ void TrackState::Start(void)
 	}
 
-	try
+	try 
```
trailing white space

DomSyna commented on fragmentcollector_hls.cpp:
```diff
@@ -5073,17 +5071,24 @@ void TrackState::Start(void)
 		playContext->reset();
 	}
-	assert(!fragmentCollectorThreadStarted);
+
```
Maybe log an err/warn where it would have asserted ? Unless its not really relevant, just assuming someone put an assert in for a reason

DomSyna commented on fragmentcollector_hls.cpp:
```diff
@@ -7055,8 +7060,8 @@ void TrackState::SwitchAudioTrack()
-                // Relative position in playlist
-                double oldPosInPlaylist = GetCompletionTimeForFragment(this, oldMediaSequenceNumber).inSeconds();
+				// Relative position in playlist
+				double oldPosInPlaylist = GetCompletionTimeForFragment(this, oldMediaSequenceNumber).inSeconds();
```
existing code, but strange alignment

DomSyna commented on fragmentcollector_hls.h:
```diff
@@ -482,8 +482,8 @@ class TrackState : public MediaTrack
 		/**
-                 * @brief Get byteRangeLength and byteRangeOffset from fragmentInfo.
-                 */
+				 * @brief Get byteRangeLength and byteRangeOffset from fragmentInfo.
```
Existing code, but bad alignment

DomSyna commented on fragmentcollector_hls.h:
```diff
@@ -1007,10 +1006,10 @@ class StreamAbstractionAAMP_HLS : public StreamAbstractionAAMP
-                 *
-                 * @return void
-                 ***************************************************************************/
+				 * @fn RefreshTrack
```
alignment

DomSyna commented on fragmentcollector_hls.h:
```diff
@@ -1026,10 +1025,10 @@ class StreamAbstractionAAMP_HLS : public StreamAbstractionAAMP
-                 * @param selectedTextTrack Current PreferredTextTrack Info
-                 * @return bool
-                 ***************************************************************************/
+				 * @fn SelectPreferredTextTrack
```
again

DomSyna reviewed (commented):
Some inherited white space / alignment issues to sort out, other than that looks fine

p-bond committed:
Tidy formatting

p-bond committed:
Formatting

p-bond committed:
Fix l1 test

p-bond committed:
Fix L1 test

p-bond committed:
Review comments.

p-bond commented on test/utests/fakes/FakePrivateInstanceAAMP.cpp:
```diff
@@ -316,6 +316,8 @@ void PrivateInstanceAAMP::ResumeDownloads()
 void PrivateInstanceAAMP::EnableDownloads()
 {
+	mDownloadsEnabled = true;
+	mDownloadsDisabled.notify_all(); // Signal the condition variable
```
Earlier iteration of the tests did use them.
Behaviour if the L1 tests unaffected by this, but may be useful for more thorough thread testing, so leaving in.

jfagunde replied:
I would not add code in case it is useful in the future. It makes it harder to understand. Also fakes should be a simple as possible, I am not sure that using conditional variables in the fake would be a good idea. I think it would be better to remove this.

p-bond committed:
Merge branch 'dev_sprint_25_2' into feature/VPLAY-8907-guardAgainstThreadOverwrite

DomSyna commented on fragmentcollector_progressive.cpp:
```diff
@@ -223,27 +223,32 @@ void StreamAbstractionAAMP_PROGRESSIVE::Start(void)
-        fragmentCollectorThreadID = std::thread(&StreamAbstractionAAMP_PROGRESSIVE::FragmentCollector, this);
-        fragmentCollectorThreadStarted = true;
-        AAMPLOG_INFO("Thread created for FragmentCollector [%zx]", GetPrintableThreadID(fragmentCollectorThreadID));
+    	// Attempting to assign to a running thread will cause std::terminate(), not an exception
```
replace space indentation

DomSyna commented on fragmentcollector_progressive.cpp:
```diff
@@ -223,27 +223,32 @@ void StreamAbstractionAAMP_PROGRESSIVE::Start(void)
+		{
+			AAMPLOG_WARN("FragmentCollector thread already running, not creating a new one");
+		}
     }
```
should be tab indentation

DomSyna commented on fragmentcollector_progressive.cpp:
```diff
@@ -223,27 +223,32 @@ void StreamAbstractionAAMP_PROGRESSIVE::Start(void)
-    if(fragmentCollectorThreadStarted)
+    if(fragmentCollectorThreadID.joinable())
     {
         aamp->DisableDownloads();
```
should be tab indentation

DomSyna commented on test/utests/tests/fragmentcollector_mpd/fragmentcollector_mpd1.cpp:
```diff
@@ -335,3 +366,40 @@ TEST_F(MpdTests, FindPositionInTimeline2)
+
+TEST_F(MpdTests, testRepeatedStartLocalTSB)
+{
+	// InitializeMPD("");
```
forgot to delete line ?

DomSyna commented on test/utests/tests/fragmentcollector_mpd/fragmentcollector_mpd1.cpp:
```diff
@@ -335,3 +366,40 @@ TEST_F(MpdTests, FindPositionInTimeline2)
+
+TEST_F(MpdTests, testRepeatedStartNotLocalTSB)
+{
+	// InitializeMPD("");
```
delete ? serves no purpose other than clutter

p-bond replied:
Deleted several times now...

DomSyna commented on test/utests/tests/fragmentcollector_progressiveTests/fragmentcollector_progressiveTests.cpp:
```diff
@@ -23,32 +23,82 @@
+        // Loop that does nothing but checks for exit condition
+        while( !threadDone)
+        {
+            // Sleep 100ms
```
pointless comment

p-bond replied:
AI

DomSyna commented on test/utests/tests/fragmentcollector_progressiveTests/fragmentcollector_progressiveTests.cpp:
```diff
@@ -23,32 +23,82 @@
     void SetUp() override
     {  
-        PrivateInstanceAAMP* aamp = new PrivateInstanceAAMP();
         aamp = new PrivateInstanceAAMP();
```
should be tab indentation

DomSyna commented on test/utests/tests/fragmentcollector_progressiveTests/fragmentcollector_progressiveTests.cpp:
```diff
@@ -23,32 +23,82 @@
+{
+    double seek_pos = 0.0;  // Provide the desired seek_pos value
+    float rate = 1.0;       // Provide the desired rate value
+
```
should be tab indentation

DomSyna reviewed (commented):
trivial stuff

p-bond committed:
WS

p-bond committed:
Merge branch 'dev_sprint_25_2' into feature/VPLAY-8907-guardAgainstThreadOverwrite

pstroffolino committed:
Merge branch 'dev_sprint_25_2' into feature/VPLAY-8907-guardAgainstThreadOverwrite
